fix estimatedWrittenBytes bug

zombee0 opened the pull request:
Signed-off-by: zombee0 <flylucas_10@163.com>

<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

We can assume that there were originally two writers and now there are three writers, so the load of each old writer is reduced by 1/3, and the load of the new writer is increased by 2/3.
before: writer0->100, writer1->100;
after:writer0->66, writer1->66; writer2->66;

<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues



<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java:
```diff
@@ -280,7 +280,7 @@ private boolean rebalancePartition(WriterId from, WriterId to)
                     if (partitionInfo.getWriterCount() <= numberOfWriters && estimatedPartitionWrittenBytes >= writerMinSize) {
```
Could you add description to commit message?

sopel39 replied:
nit: we probably should update estimated written size for all writers that were writing to a given partition as you did in https://github.com/trinodb/trino/pull/15892/files#diff-075c4ea1c37a95efc0ba3c0e3a4472d83e92a0c6034257edc76e1ee9372520dc.

But I think we can merge without that improvement too

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java:
```diff
@@ -292,13 +292,19 @@ private boolean rebalancePartition(WriterId from, WriterId to)
             return false;
         }
 
-        private void updateWriterEstimatedWrittenBytes(WriterId from, WriterId to, long estimatedPartitionWrittenBytesSinceLastRebalance, int newWriterCount)
```
Could you add a test to `TestUniformPartitionRebalancer`?

sopel39 reviewed (approved):
lgtm % comments

zombee0 commented on core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java:
```diff
@@ -292,13 +292,19 @@ private boolean rebalancePartition(WriterId from, WriterId to)
             // writer as well as increase the priority of min writer.
+            int newWriterCount = partitionInfo.getWriterCount();
             int oldWriterCount = newWriterCount - 1;
-            writerEstimatedWrittenBytes[from.id] -= (estimatedPartitionWrittenBytesSinceLastRebalance * oldWriterCount) / newWriterCount;
```
I made a new commit, I think it's more accurate. PLTA

sopel39 replied:
please push the commit when you have it.

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java:
```diff
@@ -292,13 +292,20 @@ private boolean rebalancePartition(WriterId from, WriterId to)
-            writerEstimatedWrittenBytes[to.id] += (estimatedPartitionWrittenBytesSinceLastRebalance * oldWriterCount) / newWriterCount;
+            for (int writer : partitionInfo.getWriterIds()) {
+                if (writer != to.id) {
+                    writerEstimatedWrittenBytes[writer] -= estimatedPartitionWrittenBytesSinceLastRebalance / newWriterCount;
```
nit: technically we probably should compute averaged `estimatedPartitionWrittenBytesSinceLastRebalance` from all writers
OR
compute `estimatedPartitionWrittenBytesSinceLastRebalance` per writer.

But I think it's fine for now

zombee0 replied:
I agree with you, it’s just an estimation，maybe there is not too much need of accuracy.

sopel39 replied:
Ideally `TableWriter` operator provides physical written bytes per partition, so no estimation is needed. This would require moving of `pageIndexer` logic to trino-main from connectors, which is much more involving change.

sopel39 commented:
@zombee0 how did you encountered this issue?

zombee0 commented:
> @zombee0 how did you encountered this issue?

I am teaching on how to deal with dara skew and trino give me great inspiration

arhimondr commented:
@zombee0 

Looks reasonable to me, but @gaurav8297 would be a much better reviewer for this PR.

Could you please capitalize the commit message or rewrite it, for example: `Improve written bytes estimation in UniformPartitionRebalancer`. (you can get familiar with our commit messages guidelines here: https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#commits-and-pull-requests)

gaurav8297 commented on core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java:
```diff
@@ -162,17 +162,15 @@ private synchronized void rebalancePartitions(List<Long> writerPhysicalWrittenBy
                 if (context.rebalancePartition(maxWriter, minSkewedWriter)) {
-                    maxWriters.addOrUpdate(maxWriter, context.getWriterEstimatedWrittenBytes(maxWriter));
-                    minWriters.addOrUpdate(maxWriter, Long.MAX_VALUE - context.getWriterEstimatedWrittenBytes(maxWriter));
+                    for (Integer writer: context.rebalancedAffectedWriters) {
```
Please make it a getter `context.getRebalancedAffectedWriters()`

gaurav8297 replied:
> Do you mean update all the writers although some of them have no change?
Yes

> more compution spending
I don't think it matters a lot here since `UniformPartitionRebalancer#rebalancePartitions(..)` is only called after every 50MBs has been written, so it's not that frequent. Additionally, in most cases, the total number of writers will be 32 or a little more.

gaurav8297 replied:
Why did you remove this? @zombee0

zombee0 commented:
> Overall, It looks good. Added a few minor comments. Also, please add tests and fix failures in CI.

hi, I have added an unit test, but I don't know how to run the specific ut, is there any documents  or tutorials?

gaurav8297 commented:
> hi, I have added an unit test, but I don't know how to run the specific ut, is there any documents or tutorials?

Are you using Intellij Idea? You should be able to run tests directly from there if you have docker installed.

gaurav8297 commented:
Also please run `./mvnw validate` to resolve issues related to `ci / maven-checks (17) (pull_request)`

zombee0 commented:
> > hi, I have added an unit test, but I don't know how to run the specific ut, is there any documents or tutorials?
> 
> Are you using Intellij Idea? You should be able to run tests directly from there if you have docker installed.

I am using IntelliJ IDEA, where is the docker image?

gaurav8297 commented:
> I am using IntelliJ IDEA, where is the docker image?

So, for this test `TestUniformPartitionRebalancer` there's no need for docker also. Docker is usually required in connector tests and you don't have to pull anything even in that case, the test will automatically create containers if required.

Just build the project and you should be able to test inside `TestUniformPartitionRebalancer` directly from intellij

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java:
```diff
@@ -162,17 +162,15 @@ private synchronized void rebalancePartitions(List<Long> writerPhysicalWrittenBy
             for (WriterId minSkewedWriter : minSkewedWriters) {
                 // There's no need to add the maxWriter back to priority queues if no partition rebalancing happened
                 if (context.rebalancePartition(maxWriter, minSkewedWriter)) {
```
Could you just return `rebalancedAffectedWriters` from `rebalancePartition`? When it's empty it means rebalance didn't happen.

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java:
```diff
@@ -162,17 +162,15 @@ private synchronized void rebalancePartitions(List<Long> writerPhysicalWrittenBy
             }
 
-            // Add all the min skewed writers back to the minWriters queue with updated priorities
-            for (WriterId minSkewedWriter : minSkewedWriters) {
```
I think this is still needed. `findSkewedMinWriters` will pool min writers, but `rebalancePartition` might not rebalance anything (e.g. because partition is already written by multiple writers). Hence, `min skewed` writers need to be put back into queue for next `maxWriter`

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java:
```diff
@@ -280,7 +279,8 @@ private boolean rebalancePartition(WriterId from, WriterId to)
                         rebalancedPartitions.add(partitionToRebalance.id);
-                        updateWriterEstimatedWrittenBytes(from, to, estimatedPartitionWrittenBytesSinceLastRebalance, partitionInfo.getWriterCount());
+                        rebalancedAffectedWriters.clear();
+                        updateAffectedWriterandEstimatedWrittenBytes(to, estimatedPartitionWrittenBytesSinceLastRebalance, partitionInfo);
```
```
return updateAffectedWriterandEstimatedWrittenBytes(..)
```
Make `updateAffectedWriterandEstimatedWrittenBytes` return `Set<Integer>` of rebalanced writers

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/exchange/TestUniformPartitionRebalancer.java:
```diff
@@ -107,6 +107,105 @@ public void testRebalanceWithWriterSkewness()
+    @Test
+    public void testRebalanceEffectionOnContext()
+    {
+        AtomicLong physicalWrittenBytesForWriter0 = new AtomicLong(0);
```
please add comments what particular section of test does, e.g:
```
// do X
...

// check Y
...
```

gaurav8297 commented:
Test failures seems unrelated

gaurav8297 commented:
Why did you remove the change where we update the priority of `maxWriters`? https://github.com/trinodb/trino/pull/15892#discussion_r1095580630

zombee0 commented:
> Why did you remove the change where we update the priority of `maxWriters`? [#15892 (comment)](https://github.com/trinodb/trino/pull/15892#discussion_r1095580630)

<img width="942" alt="截屏2023-02-07 14 26 56" src="https://user-images.githubusercontent.com/28446271/217165477-d99b4517-44d3-47d9-a23a-f88e2332953b.png">
please review the whole file, I had just make several commits together, I updated the maxwriter and  minskewedwriters' priority  both in maxWriters and minWriters

zombee0 commented:
> Why did you remove the change where we update the priority of `maxWriters`? [#15892 (comment)](https://github.com/trinodb/trino/pull/15892#discussion_r1095580630)

I think there is no need to return the affected writers, we just update the estimated writenbytes and we use this in context update all the maxwtier and minskewwriters that must contain all affected writers.

zombee0 commented:
@sopel39 @gaurav8297 I think it's ok to merge, PTAL

gaurav8297 commented on core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java:
```diff
@@ -256,9 +259,10 @@ private RebalanceContext(List<Long> writerPhysicalWrittenBytes, Long2LongMap par
+        private List<WriterId> rebalancePartition(WriterId from, WriterId to)
         {
             IndexedPriorityQueue<PartitionIdWithRowCount> maxPartitions = writerMaxPartitions.get(from.id);
+            List<WriterId> affectedWriters = new ArrayList<>();
```
Please use ImmutableList.Builder here

gaurav8297 commented on core/trino-main/src/test/java/io/trino/operator/exchange/TestUniformPartitionRebalancer.java:
```diff
@@ -107,6 +107,120 @@ public void testRebalanceWithWriterSkewness()
     }
 
+    @Test
+    public void testRebalanceEffectionOnContext()
```
This test looks pretty generic. Please add a test which verifies the use case. For instance, due to a particular partition rebalance, a few writers priorities are affected and they go down in the priority queue such that now they are not considered skewed in the next `while` iteration.

gaurav8297 reviewed (approved):
Some minor comments

gaurav8297 commented:
@zombee0 can you merge all commits into a single one? The first one that describes the use case.

zombee0 commented:
> @zombee0 can you merge all commits into a single one?

sure, will do it.

gaurav8297 commented:
CI hit % https://github.com/trinodb/trino/issues/12385

zombee0 commented:
> CI hit % #12385

Hey @gaurav8297 what should I do? wait and rebase?

sopel39 commented:
Thanks @zombee0 

I don't think it's user visible, please correct me if I'm wrong

zombee0 commented:
> Thanks @zombee0
> 
> I don't think it's user visible, please correct me if I'm wrong

@sopel39 yes, it's not user visible.
